runtime-modules/common/src/costs.rs

```diff
@@ -0,0 +1,135 @@
+    }
+
+    let locks = balances::Pallet::<T>::locks(account);
+    let free_balance = balances::Pallet::<T>::free_balance(account);
+    let total_balance = balances::Pallet::<T>::total_balance(account);
+
+    // Balance suitable for covering fees is equal to free_balnce
+    // minus the highest lock excluding the invitation lock
+    let unusable_free_balance = locks
+        .iter()
+        .map(|lock| {
+            if lock.id == InvitedMemberLockId::get() {
+                Zero::zero()
+            } else {
+                lock.amount
+            }
+        })
+        .max()
+        .unwrap_or_else(Zero::zero);
+    let usable_funds = free_balance.saturating_sub(unusable_free_balance);
+
+    usable_funds >= amount && total_balance >= amount.saturating_add(T::ExistentialDeposit::get())
+}
+
+// Pay a certain fee associated with a transaction (for example: a bloat bond, data size fee)
+// Fees are assumed to be payable w/ invitation-locked balance.
+// Returns the part of amount that was covered with invitation-locked balance (possibly 0).
+pub fn pay_fee<T: frame_system::Config + balances::Config>(
+    payer: &T::AccountId,
+    reciever: Option<&T::AccountId>,
+    amount: T::Balance,
+) -> Result<T::Balance, DispatchError> {
+    if amount == Zero::zero() {
+        return Ok(Zero::zero());
+    }
+
+    let account = T::AccountStore::get(payer);
+
+    ensure!(
+        has_sufficient_balance_for_fees::<T>(payer, amount),
+        DispatchError::Other("pay_fee: Insufficient balance")
+    );
+
+    // Calculate payer's new free balance
+    let new_free_balance = account
+        .free
+        .checked_sub(&amount)
+        .ok_or(DispatchError::Other("pay_fee: Free balance underflow"))?;
+
+    // Set payer's new free_balance
+    let withdraw_imbalance = balances::Pallet::<T>::make_free_balance_be(payer, new_free_balance);
+
+    // Check balance update result
+    // Make sure that the expected amount was withdrawn from payer's free_balance
+    match withdraw_imbalance {
+        SignedImbalance::Negative(negative_imbalance) if negative_imbalance.peek() != amount => {
+            Err(DispatchError::Other(
+                "pay_cost: Unexpected withdrawal imbalance - Amount does not match cost.amount",
+            ))
+        }
+        SignedImbalance::Positive(_) => Err(DispatchError::Other(
+            "pay_cost: Unexpected withdrawal imbalance - Positive",
+        )),
+        _ => Ok(()),
+    }?;
+
+    if let Some(reciever) = reciever {
+        // Deposit funds into reciever's account
+        let deposit_imbalance = balances::Pallet::<T>::deposit_creating(reciever, amount);
+
+        // Check deposit result
+        ensure!(
+            deposit_imbalance.peek() == amount,
+            DispatchError::Other("pay_cost: Unexpected deposit imbalance")
+        );
+    }
+
+    // locked_balance_used = min(account.frozen(Reasons::All).saturating_sub(new_free_balance), amount)
+    let locked_balance_used = account
```

Comment:
You're right, it wouldn't really matter whether it's `Reasons:All` or `Reasons:Fee`, because `has_sufficient_balance_for_payment` check that's done before will prevent eating into any other lock than the invitation lock.
This was simplified on Ignazio's request however: https://github.com/Joystream/joystream/pull/4164/commits/f095684787b5b46b3c360a3d6cb34ea033a4a345